packages/keystone/fields/CustomFile/Implementation.js

```diff
@@ -0,0 +1,223 @@
+                }, context)
+            }
+
+            if (!fileMeta.modelNames.includes(listKey)) {
+                throw new GQLError({
+                    code: 'FORBIDDEN',
+                    type: 'ACCESS_DENIED',
+                    variable: [this.path],
+                    message: 'Owner of this file restrict connection to this model',
+                }, context)
+            }
+        }
+
+        return await super.validateInput(props)
+    }
+
+    async resolveInput ({ resolvedData, existingItem, context, listKey }) {
+        const input = resolvedData[this.path]
+
+        // === New flow (signature): don't write a file yet; just mark this request ===
+        if (get(input, 'signature')) {
+            let fileMeta
+
+            try {
+                fileMeta = jwt.verify(input.signature, this._fileSecret, { algorithms: ['HS256'] })
+            } catch (err) {
+                throw new GQLError({
+                    code: 'BAD_USER_INPUT',
+                    type: 'WRONG_SIGNATURE',
+                    variable: [this.path],
+                    message: 'Signature is wrong or expired',
+                })
+            }
+
+            validateFileUploadSignature(fileMeta)
+
+            // keep per-request state so afterChange knows to run the webhook
+            if (!context._fileNewFlow) context._fileNewFlow = {}
+
+            const key = `${listKey}.${this.path}`
+            context._fileNewFlow[key] = {
+                signature: input.signature,
+                userId: context.authedItem?.id || null,
+                listKey,
+            }
+            return null // field stays null for the first write
+        }
+
+        if (this._strictMode) {
+            throw new GQLError({
+                code: 'BAD_USER_INPUT',
+                type: 'LEGACY_FLOW_RESTRICTED',
+                variable: [this.path],
+                message: 'You are unable to use graphql upload flow',
+            })
+        }
+
+        // === Legacy flow (Upload stream): delegate to base ===
+        return super.resolveInput({ resolvedData, existingItem })
+    }
+
+    async beforeChange ({ resolvedData, context, listKey }) {
+        const key = `${listKey}.${this.path}`
+        const marker = context._fileNewFlow && context._fileNewFlow[key]
+        if (!marker) return
+
+        const payload = {
+            id: resolvedData.id,
+            modelName: listKey,
+            signature: context._fileNewFlow[key].signature,
+            fileClientId: this._fileClientId,
+            dv: 1, sender: { dv: 1, fingerprint: 'file-attach-handler' },
+        }
+
+        let attachResult
+
+        try {
+            const res = await fetch(this._fileServiceUrl, {
+                method: 'POST',
+                headers: { 'Content-Type': 'application/json' },
+                body: JSON.stringify(payload),
+            })
+
+            if (!res.ok) {
+                return
```

Comment:
Why just return here (and below in afterChange)? Can we throw an error to user?

Reply:
That's because of tricky api miscommunications. So from that part we had an graphql like syntax, but new API is REST-like. 
Anyway now I catch all of the errors, convert it back to GQLError factory and finally throw to user. 
Also if there is something unexpected - I throw that too. 

Maybe also I want to add some logs to the unexpected part, so developers may understand what's going bad here